### ISSUE TITLE
Inconsistent frame/elapsed time for camera sensor data in 0.9.10-2

### ISSUE THREAD

**Issue by analog-cbarber** [carla-simulator/carla#3391]
I am seeing inconsistent frame numbers and elapsed time values using the latest 0.9.10-2 debian release on Ubuntu.

I have scenario where I am placing random pedestrians in each frame, and viewing them with three co-located cameras: rgb, depth and segmentation. I am using synchronous mode with a fixed time delta of .1s (10fps). This worked ok with 0.9.10-1,
but with the latest release I am sometimes getting consecutive records from each sensor that have the same frame number or same elapsed time as the previous frame and if you look at the images, they don't line up consistently. So I basically cannot trust the frame and timestamp labels I am getting from Carla.

Here are a few frames from the segmentation and rgb cameras. In this case, the segmentation camera frame numbers are one less than the corresponding rgb frames and the elapsed times are either one behind or are just wrong. But there is no consistent pattern even within a single episode.

![CarlaBug](https://user-images.githubusercontent.com/7408243/94966563-7b64ae00-04cb-11eb-85fa-426c0af21bc0.png)

I don't have a simple script to reproduce this right now, but I can write one if you need it.

Is this a known issue?

**Comment by analog-cbarber** [comment]
Here is a script to reproduce at least some of the bad behavior.

Using this script, I have not observed a total mismatch of images with the same frame #'s as was shown in the above example, but it does usually show duplicate frames and mismatches with frames with the same elapsed timestamp. I suspect the reason my own code is showing worse effect is due to some difference in timing.

Unless there is some workaround for this, I don't think I can do any sensor fusion with the data from this version of Carla.

```python
"""
Script to reproduce Carla bug 3391
"""

import argparse
import csv
import os
import re
import subprocess
import sys
import time
from abc import ABC, abstractmethod
from glob import glob
from queue import Queue, Empty
from shutil import rmtree
from threading import Thread
from typing import cast

carla_root = os.getenv('CARLA_PYTHON_ROOT', '/opt/carla-simulator')
assert os.path.isdir(carla_root)
glob_expr = os.path.join(carla_root, 'PythonAPI', 'carla', 'dist', 'carla-*-py3.7-*.egg')
print(glob_expr)
carla_egg = glob(glob_expr)
sys.path.append(carla_egg[0])

import carla
import numpy as np

class CarlaSession:
    server_process = None
    client = None

    def __init__(self):
        pass

    def start(self):
        self.server_process = subprocess.Popen(
            [os.path.join(carla_root, 'CarlaUE4','Binaries', 'Linux', 'CarlaUE4-Linux-Shipping'), 'CarlaUE4'],
            stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL
        )
        time.sleep(5)
        self.client = carla.Client('localhost', 2000)

    def stop(self):
        proc = self.server_process
        proc.terminate()
        for i in range(20):
            try:
                proc.wait(1)
            except (TimeoutError, subprocess.TimeoutExpired):
                print(i)
                proc.kill()

    def __enter__(self):
        self.start()
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.stop()


class Camera(ABC):
    name:str
    input_queue:Queue
    thread:Thread
    actor:carla.Sensor
    timeout:float = 10.0
    outdir:str = ''
    frames:int = 10

    def __init__(self, name, *,
                 world:carla.World,
                 pose:carla.Transform
                 ):
        self.name = name
        self.input_queue = Queue(maxsize=100)
        bp = world.get_blueprint_library().find(f'sensor.camera.{name}')
        bp.set_attribute('fov', '120')
        bp.set_attribute('image_size_x', '1000')
        bp.set_attribute('image_size_y', '1000')
        bp.set_attribute('role_name', name)
        self.actor = world.spawn_actor(bp, pose)
        self.thread = Thread(
            target = self.process_queue,
            name = f"{name}_thread",
            daemon = True
        )

    @property
    def frames_file(self) -> str:
        return os.path.join(self.outdir, f'{self.name}-frames.csv')

    def data_file(self, offset:int) -> str:
        return os.path.join(self.outdir, f'{self.name}-{int(offset)}.npy')

    def process_queue(self):
        q = self.input_queue
        try:
            count = 0
            prev_frame = -1
            prev_elapsed = -1.0
            with open(self.frames_file, 'w', newline='') as f:
                csvout = csv.writer(f, dialect='unix', quoting=csv.QUOTE_MINIMAL)
                csvout.writerow(['n','frame','elapsed'])
                while count < self.frames:
                    entry:carla.Image = q.get(timeout = self.timeout)
                    a = self.to_array(entry)
                    np.save(self.data_file(count), a)
                    csvout.writerow([count, entry.frame, entry.timestamp])
                    if entry.frame == prev_frame:
                        print(f'{self.name} {count} has same frame # as previous # {prev_frame}',
                              file = sys.stderr)
                    if entry.timestamp == prev_elapsed:
                        print(f'{self.name} {count} has same elapsed time as previous {prev_elapsed:.2f}',
                              file = sys.stderr)
                    prev_frame = entry.frame
                    prev_elapsed = entry.timestamp
                    count += 1
        except Empty:
            pass

    def read_frame_data(self, strip_duplicates=False) -> np.ndarray:
        with open(self.frames_file, newline='') as f:
            csvin = csv.reader(f, dialect='unix')
            rows = list(csvin)

        data = np.array(rows[1:], dtype=float)

        if strip_duplicates:
            # strip any rows that contain an entry with a non-unique frame or elapsed value
            frames = data[:,1]
            _, idx, count = np.unique(frames, return_index=True, return_counts=True)
            uframeidx = idx[count == 1]
            elapsed = data[:,2]
            _, idx, count = np.unique(elapsed, return_index=True, return_counts=True)
            uelapsedidx = idx[count == 1]
            idx = sorted(set(uframeidx).intersection(uelapsedidx))
            data = data[idx]

        return data

    def start_listening(self):
        self.actor.listen(self.enqueue_input)

    def enqueue_input(self, image:carla.Image):
        self.input_queue.put(image)

    @abstractmethod
    def to_array(self, image:carla.Image) -> np.ndarray:
        pass


class RGBCamera(Camera):
    def __init__(self, *, world, pose):
        super().__init__('rgb', world=world, pose=pose)

    def to_array(self, image:carla.Image) -> np.ndarray:
        bgra = np.frombuffer(image.raw_data, dtype=np.uint8).reshape(image.height, image.width, -1)
        return np.flip(bgra, axis=-1)[:,:,1:]


class DepthCamera(Camera):
    def __init__(self, *, world, pose):
        super().__init__('depth', world=world, pose=pose)

    def to_array(self, image:carla.Image) -> np.ndarray:
        rawdepth = np.frombuffer(image.raw_data, dtype='>u4').reshape(image.height, image.width)
        # BGRA form where BGR holds the depth where max value is 1000 meters
        return cast(np.ndarray, np.multiply(rawdepth, 1000.0 / 0xFFFFFF00, dtype=np.float32))


class SegCamera(Camera):
    def __init__(self, *, world, pose):
        super().__init__('semantic_segmentation', world=world, pose=pose)

    def to_array(self, image:carla.Image) -> np.ndarray:
        bgra = np.frombuffer(image.raw_data, dtype=np.uint8).reshape(image.height, image.width, -1)
        # initially in BGRA format, only the red channel is used
        return bgra[:,:,2].copy()

def main(argv=None):
    basedir = os.getcwd()
    out_prefix = os.path.splitext(os.path.basename(__file__))[0]
    prev_dirs = glob(os.path.join(basedir, out_prefix + '-*'))
    highest_counter = 0
    for path in prev_dirs:
        m = re.search('-(\d+)', path)
        if m:
            highest_counter = max(highest_counter, int(m.group(1)))
    outdir = os.path.join(basedir, out_prefix + f'-{highest_counter+1}')

    parser = argparse.ArgumentParser(
        description = "Reproduce Carla issue #3391"
    )
    parser.add_argument(
        '--frames', metavar='<n>', type=int, default=10,
        help="Number of frames to generate"
    )
    parser.add_argument(
        '--out', metavar='<dir>', default=outdir,
        help = 'Output directory (%(default)s).'
    )
    parser.add_argument(
        '--clean', action='store_true',
        help = 'Remove existing output directories'
    )

    parsed = parser.parse_args(argv)

    frames = int(parsed.frames)

    if parsed.clean:
        for path in prev_dirs:
            rmtree(path, ignore_errors=True)
        if parsed.out == outdir:
            parsed.out = os.path.join(basedir, f'{out_prefix}-1')

    outdir = parsed.out
    os.makedirs(outdir, exist_ok=False)

    with CarlaSession() as session:
        client = session.client
        world = client.load_world('Town01')
        world.apply_settings(
            carla.WorldSettings(
                synchronous_mode = True,
                fixed_delta_seconds = 0.1
            )
        )
        location = carla.Location(68.133, 192.315, 3)
        pose = carla.Transform(location, carla.Rotation(-90,0,0))
        world.get_spectator().set_transform(pose)

        world.set_weather(carla.WeatherParameters.ClearNoon)

        bpl = world.get_blueprint_library()

        cameras = tuple(
            Cam(world=world, pose=pose)
            for Cam in (RGBCamera, DepthCamera, SegCamera)
        )
        rgbcam, depthcam, segcam = cameras

        for camera in cameras:
            camera.outdir = outdir
            camera.frames = frames
            camera.thread.start()


        # tick once so that sensors will have locations
        world.tick()
        # world.tick()

        # HACK: not sure why this sleep is necessary, but otherwise
        # we end up getting an empty initial frame.
        time.sleep(1)

        for camera in cameras:
            camera.start_listening()

        # Generate a frame with no pedestrians
        world.tick()

        frame_data = []

        for _ in range(frames):
            people_bps = list(bpl.filter('pedestrian'))
            n_people = np.random.randint(1, len(people_bps))
            people_bps = np.random.choice(people_bps, n_people, replace=False)
            people = []
            for bp in people_bps:
                person_pose = carla.Transform(
                    carla.Location(
                        x=location.x + np.random.uniform(-3.0, 3.0),
                        y=location.y + np.random.uniform(-3.0, 3.0),
                        z=.1,
                    ),
                    carla.Rotation(
                        yaw=np.random.uniform(-180.0, 180.0)
                    )
                )
                try:
                    people.append(world.spawn_actor(bp, person_pose))
                except RuntimeError:
                    pass # probably a collision

            world.tick()

            for person in people:
                person.destroy()

        for camera in cameras:
            camera.thread.join()

        # Read CSV files
        rgb_frames = rgbcam.read_frame_data(strip_duplicates=False)
        seg_frames = segcam.read_frame_data(strip_duplicates=False)
        depth_frames = segcam.read_frame_data(strip_duplicates=False)

        rgb0 = np.load(rgbcam.data_file(0))
        seg0 = np.load(segcam.data_file(0))
        depth0 = np.load(depthcam.data_file(0))

        PEDESTRIAN_FLAG = 4
        # there should be no pedestrians in first frame
        assert np.all(seg0 != PEDESTRIAN_FLAG)

        for row in seg_frames[1:]:
            n = int(row[0])
            frame = int(row[1])
            elapsed = float(row[2])
            segn = np.load(segcam.data_file(n))
            pedmask = segn == PEDESTRIAN_FLAG
            if not np.any(pedmask):
                continue # no pedestrians in view

            # find corresponding depth frame
            for row in depth_frames:
                depth_frame = int(row[1])
                depth_elapsed = float(row[2])
                if depth_elapsed == elapsed:
                # if depth_frame == frame:
                    dn = int(row[0])
                    depthn = np.load(depthcam.data_file(dn))
                    depthn[pedmask] = depth0[pedmask]
                    # the pixels without pedestrians in them should have the same values
                    # really they should be the same, but we will allow a fudge factor
                    norm = np.linalg.norm(depthn - depth0)
                    isclose = norm < 10
                    # if isclose:
                    #     print(f'Depth and seg at frame {frame} are ok')
                    if not isclose:
                        print(f'Depth {dn} and seg {n} at frame {frame} {elapsed:.2f}s are mismatched',
                              file=sys.stderr)
                    break

if __name__ == '__main__':
    main()
```

**Comment by thytless** [comment]
Hello! Did your problem solved in the latest release 0.9.10.1?

**Comment by PTorr** [comment]
I'm not sure, but I think your problem is that you use wold.tick() multiple times there.

It should be use once for each iteration of the synchronous mode.
